includes/Core/Storage/Setting_With_Owned_Keys_Trait.php

```diff
@@ -36,29 +37,53 @@ abstract public function get_owned_keys();
 	 * @since 1.16.0
 	 */
 	protected function register_owned_keys() {
-		$keys = $this->get_owned_keys();
+		$call_when_owned_settings_changed = function( $settings, $old_settings, $callback ) {
+			if ( ! current_user_can( Permissions::MANAGE_OPTIONS ) ) {
+				return;
+			}
+
+			$keys = $this->get_owned_keys();
+			foreach ( $keys as $key ) {
+				if ( isset( $settings[ $key ], $old_settings[ $key ] ) && $settings[ $key ] !== $old_settings[ $key ] ) {
+					call_user_func( $callback );
+					break;
+				}
+			}
+		};
 
 		add_action(
 			'add_option_' . static::OPTION,
-			function ( $option, $value ) use ( $keys ) {
-				if ( is_array( $value ) && count( array_intersect( array_keys( $value ), $keys ) ) > 0 && current_user_can( Permissions::MANAGE_OPTIONS ) ) {
-					$this->merge( array( 'ownerID' => get_current_user_id() ) );
+			function ( $option, $value ) use ( $call_when_owned_settings_changed ) {
+				if ( ! is_array( $value ) || ! $this instanceof Setting ) {
+					return;
+				}
+
+				$defaults = $this->get_default();
+				if ( ! is_array( $defaults ) ) {
+					return;
 				}
+
+				$call_when_owned_settings_changed(
+					$value,
+					$defaults,
+					array( $this, 'merge_initial_owner_id' )
+				);
 			},
 			10,
 			2
 		);
 
 		add_filter(
 			'pre_update_option_' . static::OPTION,
-			function( $value, $old_value ) use ( $keys ) {
+			function ( $value, $old_value ) use ( $call_when_owned_settings_changed ) {
 				if ( is_array( $value ) && is_array( $old_value ) ) {
-					foreach ( $keys as $key ) {
-						if ( isset( $value[ $key ], $old_value[ $key ] ) && $value[ $key ] !== $old_value[ $key ] && current_user_can( Permissions::MANAGE_OPTIONS ) ) {
-							$value['ownerID'] = get_current_user_id();
-							break;
+					$call_when_owned_settings_changed(
+						$value,
+						$old_value,
+						function() use ( &$value ) {
+							$value = $this->filter_owner_id_for_updated_settings( $value );
```

Comment:
Passing variables by reference is possible but very unwelcome in PHP as far as I remember because it is very easy to break it in the future when someone removes the `&` by accident during refactoring or another maintenance. Knowing that we have engineers with different level of PHP knowledge, I would not recommend to use it.

Reply:
Ok, fair enough, let's leave it then...

I think we could still change the method name as it doesn't really seem like a filter to me... what do you think about `update_owner_id_in_settings()` ?

Reply:
Yes, no problem. Updated.